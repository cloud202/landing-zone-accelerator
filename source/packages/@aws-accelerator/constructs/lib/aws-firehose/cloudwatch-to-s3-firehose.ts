/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { pascalCase } from 'change-case';
import * as path from 'path';

/**
 * Construction properties for CloudWatch to S3 replication for Kinesis Stream.
 */
export interface CloudWatchToS3FirehoseProps {
  /**
   *
   * Dynamic partitioning as JSON string array to partition records in firehose
   */
  dynamicPartitioningValue?: string;
  /**
   *
   * Source bucket object is must when source bucket name wasn't provided
   * This bucket will have all the CloudWatch Logs
   */
  bucket?: cdk.aws_s3.IBucket;
  /**
   *
   * Source bucket name is must when source bucket object wasn't provided
   * This bucket will have all the CloudWatch Logs
   */
  bucketName?: string;
  /**
   *
   * Kinesis Stream which will trigger Firehose
   */
  kinesisStream: cdk.aws_kinesis.IStream;
  /**
   *
   * KMS key to encrypt the Firehose
   */
  firehoseKmsKey: cdk.aws_kms.IKey;
  /**
   *
   * KMS key to encrypt the Lambda
   */
  lambdaKey: cdk.aws_kms.IKey;
  /**
   *
   * Home region where the log archive bucket is located.
   */
  homeRegion: string;
  /**
   *
   * KMS key to encrypt the Lambda
   */
  kinesisKmsKey: cdk.aws_kms.IKey;
}
/**
 * Class to configure CloudWatch replication on logs receiving account
 */
export class CloudWatchToS3Firehose extends Construct {
  constructor(scope: Construct, id: string, props: CloudWatchToS3FirehoseProps) {
    super(scope, id);

    let dynamicPartitioning = '';
    if (props.dynamicPartitioningValue) {
      dynamicPartitioning = props.dynamicPartitioningValue;
    }

    let logsStorageBucket: cdk.aws_s3.IBucket;
    if (props.bucket && props.bucketName) {
      throw new Error('Source bucket or source bucketName (only one property) should be defined.');
    }

    if (!props.bucket && !props.bucketName) {
      throw new Error('Source bucket or source bucketName property must be defined when using bucket replication.');
    }

    if (props.bucketName) {
      logsStorageBucket = cdk.aws_s3.Bucket.fromBucketName(this, `${pascalCase(props.bucketName)}`, props.bucketName);
    } else {
      logsStorageBucket = props.bucket!;
    }

    const firehosePrefixProcessingLambda = new cdk.aws_lambda.Function(this, 'FirehosePrefixProcessingLambda', {
      runtime: cdk.aws_lambda.Runtime.NODEJS_14_X,
      functionName: 'AWSAccelerator-FirehoseRecordsProcessor',
      code: cdk.aws_lambda.Code.fromAsset(path.join(__dirname, 'firehose-record-processing/dist')),
      handler: 'index.handler',
      memorySize: 2048,
      timeout: cdk.Duration.minutes(5),
      environmentEncryption: props.lambdaKey,
      environment: {
        DynamicS3LogPartitioningMapping: dynamicPartitioning,
      },
    });

    // Access is based on least privileged apis
    // Ref: https://docs.aws.amazon.com/firehose/latest/dev/controlling-access.html#using-iam-s3
    const firehoseAccessS3KmsLambda = new cdk.aws_iam.PolicyDocument({
      statements: [
        new cdk.aws_iam.PolicyStatement({
          actions: ['kms:Decrypt', 'kms:GenerateDataKey'],
          resources: [props.firehoseKmsKey.keyArn],
          conditions: {
            StringEquals: {
              'kms:ViaService': `s3.${props.homeRegion}.amazonaws.com`,
            },
            StringLike: {
              'kms:EncryptionContext:aws:s3:arn': `${logsStorageBucket.bucketArn}/*`,
            },
          },
        }),
        new cdk.aws_iam.PolicyStatement({
          actions: [
            's3:AbortMultipartUpload',
            's3:GetBucketLocation',
            's3:GetObject',
            's3:ListBucket',
            's3:ListBucketMultipartUploads',
            's3:PutObject',
          ],
          resources: [logsStorageBucket.bucketArn, `${logsStorageBucket.bucketArn}/*`],
        }),
        new cdk.aws_iam.PolicyStatement({
          actions: ['lambda:InvokeFunction', 'lambda:GetFunctionConfiguration'],
          resources: [
            `${firehosePrefixProcessingLambda.functionArn}:*`,
            `${firehosePrefixProcessingLambda.functionArn}`,
          ],
        }),
      ],
    });

    const firehoseServiceRole = new cdk.aws_iam.Role(this, 'FirehoseS3ServiceRole', {
      assumedBy: new cdk.aws_iam.ServicePrincipal('firehose.amazonaws.com'),
      description: 'Role used by Kinesis Firehose to place Kinesis records in the central bucket.',
      // placing inline policy as firehose needs this from get-go or there might be a few initial failures
      inlinePolicies: {
        AccessS3KmsLambda: firehoseAccessS3KmsLambda,
      },
    });

    const firehoseAccessKinesis = new cdk.aws_iam.PolicyDocument({
      statements: [
        new cdk.aws_iam.PolicyStatement({
          actions: ['kinesis:DescribeStream', 'kinesis:GetShardIterator', 'kinesis:GetRecords', 'kinesis:ListShards'],
          resources: [props.kinesisStream.streamArn],
        }),

        new cdk.aws_iam.PolicyStatement({
          actions: [
            'kms:Decrypt',
            'kms:Encrypt',
            'kms:GenerateDataKey',
            'kms:ReEncryptTo',
            'kms:GenerateDataKeyWithoutPlaintext',
            'kms:GenerateDataKeyPairWithoutPlaintext',
            'kms:GenerateDataKeyPair',
            'kms:ReEncryptFrom',
          ],
          resources: [props.kinesisKmsKey.keyArn],
        }),
      ],
    });

    const kinesisStreamRole = new cdk.aws_iam.Role(this, 'FirehoseKinesisStreamServiceRole', {
      assumedBy: new cdk.aws_iam.ServicePrincipal('firehose.amazonaws.com'),
      description: 'Role used by Kinesis Firehose to get records from Kinesis.',
      // placing inline policy as firehose needs this from get-go or there might be a few initial failures
      inlinePolicies: {
        AccessKinesis: firehoseAccessKinesis,
      },
    });

    new cdk.aws_kinesisfirehose.CfnDeliveryStream(this, 'Kinesis-Firehose-Stream-Dynamic-Partitioning', {
      deliveryStreamType: 'KinesisStreamAsSource',
      kinesisStreamSourceConfiguration: {
        kinesisStreamArn: props.kinesisStream.streamArn,
        roleArn: kinesisStreamRole.roleArn,
      },
      extendedS3DestinationConfiguration: {
        bucketArn: logsStorageBucket.bucketArn,
        bufferingHints: {
          intervalInSeconds: 60,
          sizeInMBs: 64, // Minimum with dynamic partitioning
        },
        compressionFormat: 'UNCOMPRESSED',
        roleArn: firehoseServiceRole.roleArn,
        dynamicPartitioningConfiguration: {
          enabled: true,
        },
        errorOutputPrefix: `CloudWatchLogs/processing-failed`,
        encryptionConfiguration: {
          kmsEncryptionConfig: {
            awskmsKeyArn: props.firehoseKmsKey.keyArn,
          },
        },
        prefix: '!{partitionKeyFromLambda:dynamicPrefix}',
        processingConfiguration: {
          enabled: true,
          processors: [
            {
              type: 'Lambda',
              parameters: [
                {
                  parameterName: 'LambdaArn',
                  parameterValue: firehosePrefixProcessingLambda.functionArn,
                },
                {
                  parameterName: 'NumberOfRetries',
                  parameterValue: '3',
                },
              ],
            },
          ],
        },
      },
    });
  }
}
