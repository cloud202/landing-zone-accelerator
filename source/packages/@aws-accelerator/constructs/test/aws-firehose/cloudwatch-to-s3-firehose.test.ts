/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { CloudWatchToS3Firehose } from '../../lib/aws-firehose/cloudwatch-to-s3-firehose';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(CloudWatchToS3Firehose): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

new CloudWatchToS3Firehose(stack, 'CloudWatchToS3Firehose', {
  firehoseKmsKey: new cdk.aws_kms.Key(stack, 'CustomKey', {}),
  lambdaKey: new cdk.aws_kms.Key(stack, 'CustomLambdaKey', {}),
  kinesisStream: new cdk.aws_kinesis.Stream(stack, 'CustomStream', {}),
  kinesisKmsKey: new cdk.aws_kms.Key(stack, 'CustomKinesisKey', {}),
  bucket: new cdk.aws_s3.Bucket(stack, 'CustomBucket', {}),
  dynamicPartitioningValue: '',
  homeRegion: 'someregion',
});

/**
 * CloudWatchDestination construct test
 */
describe('CloudWatchToS3Firehose', () => {
  snapShotTest(testNamePrefix, stack);
});
