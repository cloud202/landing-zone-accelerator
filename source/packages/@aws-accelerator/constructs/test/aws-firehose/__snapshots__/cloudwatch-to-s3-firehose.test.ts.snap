// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudWatchToS3Firehose Construct(CloudWatchToS3Firehose):  Snapshot Test 1`] = `
Object {
  "Conditions": Object {
    "AwsCdkKinesisEncryptedStreamsUnsupportedRegions": Object {
      "Fn::Or": Array [
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-north-1",
          ],
        },
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-northwest-1",
          ],
        },
      ],
    },
  },
  "Resources": Object {
    "CloudWatchToS3FirehoseFirehoseKinesisStreamServiceRole80E25BD8": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Role used by Kinesis Firehose to get records from Kinesis.",
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "kinesis:DescribeStream",
                    "kinesis:GetShardIterator",
                    "kinesis:GetRecords",
                    "kinesis:ListShards",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "CustomStreamE8E9158E",
                      "Arn",
                    ],
                  },
                },
                Object {
                  "Action": Array [
                    "kms:Decrypt",
                    "kms:Encrypt",
                    "kms:GenerateDataKey",
                    "kms:ReEncryptTo",
                    "kms:GenerateDataKeyWithoutPlaintext",
                    "kms:GenerateDataKeyPairWithoutPlaintext",
                    "kms:GenerateDataKeyPair",
                    "kms:ReEncryptFrom",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "CustomKinesisKey1382C574",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "AccessKinesis",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CloudWatchToS3FirehoseFirehosePrefixProcessingLambdaE5E42603": Object {
      "DependsOn": Array [
        "CloudWatchToS3FirehoseFirehosePrefixProcessingLambdaServiceRole405E5167",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "DynamicS3LogPartitioningMapping": "",
          },
        },
        "FunctionName": "AWSAccelerator-FirehoseRecordsProcessor",
        "Handler": "index.handler",
        "KmsKeyArn": Object {
          "Fn::GetAtt": Array [
            "CustomLambdaKey2287F5A9",
            "Arn",
          ],
        },
        "MemorySize": 2048,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CloudWatchToS3FirehoseFirehosePrefixProcessingLambdaServiceRole405E5167",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CloudWatchToS3FirehoseFirehosePrefixProcessingLambdaServiceRole405E5167": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CloudWatchToS3FirehoseFirehoseS3ServiceRole8EEB3338": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Role used by Kinesis Firehose to place Kinesis records in the central bucket.",
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "kms:Decrypt",
                    "kms:GenerateDataKey",
                  ],
                  "Condition": Object {
                    "StringEquals": Object {
                      "kms:ViaService": "s3.someregion.amazonaws.com",
                    },
                    "StringLike": Object {
                      "kms:EncryptionContext:aws:s3:arn": Object {
                        "Fn::Join": Array [
                          "",
                          Array [
                            Object {
                              "Fn::GetAtt": Array [
                                "CustomBucketF81FA714",
                                "Arn",
                              ],
                            },
                            "/*",
                          ],
                        ],
                      },
                    },
                  },
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "CustomKey1E6D0D07",
                      "Arn",
                    ],
                  },
                },
                Object {
                  "Action": Array [
                    "s3:AbortMultipartUpload",
                    "s3:GetBucketLocation",
                    "s3:GetObject",
                    "s3:ListBucket",
                    "s3:ListBucketMultipartUploads",
                    "s3:PutObject",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "CustomBucketF81FA714",
                        "Arn",
                      ],
                    },
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          Object {
                            "Fn::GetAtt": Array [
                              "CustomBucketF81FA714",
                              "Arn",
                            ],
                          },
                          "/*",
                        ],
                      ],
                    },
                  ],
                },
                Object {
                  "Action": Array [
                    "lambda:InvokeFunction",
                    "lambda:GetFunctionConfiguration",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          Object {
                            "Fn::GetAtt": Array [
                              "CloudWatchToS3FirehoseFirehosePrefixProcessingLambdaE5E42603",
                              "Arn",
                            ],
                          },
                          ":*",
                        ],
                      ],
                    },
                    Object {
                      "Fn::GetAtt": Array [
                        "CloudWatchToS3FirehoseFirehosePrefixProcessingLambdaE5E42603",
                        "Arn",
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "AccessS3KmsLambda",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CloudWatchToS3FirehoseKinesisFirehoseStreamDynamicPartitioning03389B0D": Object {
      "Properties": Object {
        "DeliveryStreamType": "KinesisStreamAsSource",
        "ExtendedS3DestinationConfiguration": Object {
          "BucketARN": Object {
            "Fn::GetAtt": Array [
              "CustomBucketF81FA714",
              "Arn",
            ],
          },
          "BufferingHints": Object {
            "IntervalInSeconds": 60,
            "SizeInMBs": 64,
          },
          "CompressionFormat": "UNCOMPRESSED",
          "DynamicPartitioningConfiguration": Object {
            "Enabled": true,
          },
          "EncryptionConfiguration": Object {
            "KMSEncryptionConfig": Object {
              "AWSKMSKeyARN": Object {
                "Fn::GetAtt": Array [
                  "CustomKey1E6D0D07",
                  "Arn",
                ],
              },
            },
          },
          "ErrorOutputPrefix": "CloudWatchLogs/processing-failed",
          "Prefix": "!{partitionKeyFromLambda:dynamicPrefix}",
          "ProcessingConfiguration": Object {
            "Enabled": true,
            "Processors": Array [
              Object {
                "Parameters": Array [
                  Object {
                    "ParameterName": "LambdaArn",
                    "ParameterValue": Object {
                      "Fn::GetAtt": Array [
                        "CloudWatchToS3FirehoseFirehosePrefixProcessingLambdaE5E42603",
                        "Arn",
                      ],
                    },
                  },
                  Object {
                    "ParameterName": "NumberOfRetries",
                    "ParameterValue": "3",
                  },
                ],
                "Type": "Lambda",
              },
            ],
          },
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "CloudWatchToS3FirehoseFirehoseS3ServiceRole8EEB3338",
              "Arn",
            ],
          },
        },
        "KinesisStreamSourceConfiguration": Object {
          "KinesisStreamARN": Object {
            "Fn::GetAtt": Array [
              "CustomStreamE8E9158E",
              "Arn",
            ],
          },
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "CloudWatchToS3FirehoseFirehoseKinesisStreamServiceRole80E25BD8",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::KinesisFirehose::DeliveryStream",
    },
    "CustomBucketF81FA714": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "CustomKey1E6D0D07": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "CustomKinesisKey1382C574": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "CustomLambdaKey2287F5A9": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "CustomStreamE8E9158E": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "Fn::If": Array [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            Object {
              "Ref": "AWS::NoValue",
            },
            Object {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
        "StreamModeDetails": Object {
          "StreamMode": "PROVISIONED",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
  },
}
`;
